src/main/java/io/datanerds/avropatch/serialization/CustomTypeSerializer.java

```diff
@@ -0,0 +1,148 @@
+import org.apache.avro.Schema;
+import org.apache.avro.io.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.ThreadSafe;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+
+import static io.datanerds.avropatch.schema.PrimitiveTypes.*;
+
+@ThreadSafe
+public final class CustomTypeSerializer {
+
+    private static Logger logger = LoggerFactory.getLogger(CustomTypeSerializer.class);
+
+    private final DatumWriter<Patch> writer;
+    private final DatumReader<Patch> reader;
+
+    private CustomTypeSerializer(Schema schema) {
+        Objects.nonNull(schema);
+        this.writer = AvroData.get().createDatumWriter(schema);
+        this.reader = AvroData.get().createDatumReader(schema);
+    }
+
+    public byte[] toBytes(Patch value) throws IOException {
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        Encoder binaryEncoder = EncoderFactory.get().directBinaryEncoder(outputStream, null);
+        writer.write(value, binaryEncoder);
+        return outputStream.toByteArray();
+    }
+
+    public Patch toObject(byte[] bytes) throws IOException {
+        return reader.read(null, DecoderFactory.get().binaryDecoder(bytes, null));
+    }
+
+    public static class Builder {
+        protected final List<Schema> types;
+
+        public Builder() {
+            this.types = new ArrayList<>();
+        }
+
+        public Builder nullable() {
+            types.add(NULL);
+            return this;
+        }
+
+        public Builder withPrimitives() {
+            types.addAll(Arrays.asList(BOOLEAN, DOUBLE, FLOAT, INTEGER, LONG, STRING));
+            return this;
+        }
+
+        public Builder withCustomTypes() {
+            types.addAll(Arrays.asList(BigDecimalType.SCHEMA, BigIntegerType.SCHEMA, DateType.SCHEMA, UuidType.SCHEMA));
+            return this;
+        }
+
+        public Builder with(Type type) {
+            types.add(AvroData.get().getSchema(type));
+            return this;
+        }
+
+        public Builder with(Schema schema) {
+            types.add(schema);
+            return this;
+        }
+
+        public CustomTypeSerializer build() {
+            Schema schema = makeSchema();
+            logger.debug("Created serializer for schema {}", schema);
+            return new CustomTypeSerializer(schema);
+        }
+
+        private Schema makeSchema() {
+            if (types.size() == 1) {
+                return CustomTypes.PatchType.create(types.get(0));
+            } else {
+                return CustomTypes.PatchType.create(Schema.createUnion(types));
+            }
+        }
+
+        private Builder endSubBuilder(Schema schema) {
+            types.add(schema);
+            return this;
+        }
+
+        public Builder.ArrayBuilder withArray() {
```

Comment:
How about something like `public Builder withArray(Builder.ArrayBuilder arrayBuilder)` and then go from there? Having the ArrayBuilder explicitly as a method parameter would emphasise the nesting.